src/v/cluster/rm_stm.h

```diff
@@ -59,13 +56,79 @@ namespace cluster {
+ * transactional client registers. This ensures that only one active
+ * transactional producer is active, the one with the latest epoch.
+ *
+ * Both idempotent and transactional producers are associated with a
+ * producer_state instance and the state machine tracks the list of all active
+ * producer_states producing to this partition.
+ *
+ * Notes on locking in this stm:
+ *
+ * There are two main locks in use here.
+ * The locks are always grabbed in that order.
+ * 1. (stm)_state_lock : This is a state machine wide (global) lock acquired by
+ * all operations like producers and snapshots. Producers acquire it in read
+ * (shared) mode while operations like snapshots that require a consistent view
+ * of the state machine acquire it in write (exclusive) mode.
+ *
+ * 2. producer_state_lock: This is transparently grabbed in the invocations of
+ * run_with_lock method of the producer but the scope of the lock is different
+ * for idempotent and transactional producers.
+
+ * - Idempotent requests hold the lock until the requested cannot be reordered
+ * at the raft layer (so the sequence is not violated in the physical log) and
+ * released soon after to let concurrent idempotent requests make progress.
+ *
+ * - Transactional requests hold the lock for the duration of the request and
+ * until the changes are reflected in the state machine via apply. So the scope
+ * of the lock is larger.
+ *
+ * Notes on producer_state eviction:
+ *
+ * Kafka protocol has no notion of session termination. For example, an
+ * idempotent producer may just stop producing and the broker wouldn't know
+ * whether it is ever going to come back again. This forces the implementation
+ * to have an eviction policy for producers so we do not just accumulate
+ * them forever. This is currently based on a configurable count
+ * (max_concurrent_producer_ids), enforced by the producer_state_manager
+ * at the shard level. While this state machine owns the producer_state
+ * instances, the producer_state_manager manges a shard wide LRU
+ * list of producers and notifies the stm when the limits are breached. See
+ * producer_state class for details on when a producer can be evicted.
+ *
+ * Notes on transaction expiry:
+
+ * This stm periodically checks if there is any pending transaction for
+ * expiration. The expiration kicks in the transaction is not committed/aborted
+ * within the user set transaction timeout. A producer with an active
+ * transaction cannot be evicted, so exipration ensures that with timely
+ * expiration of open transactions, the producer states are candidates for
+ * eviction.
  */
 class rm_stm final : public raft::persisted_stm<> {
 public:
     static constexpr std::string_view name = "rm_stm";
 
-    using producers_t
-      = mt::map_t<absl::btree_map, model::producer_identity, tx::producer_ptr>;
+    using producers_t = absl::btree_map<model::producer_id, tx::producer_ptr>;
```

Comment:
It is unclear to me how we handle producer with bupped `epoch`. It seems that we only forget about producer when all transactions are aborted. The producer with new epoch will always be fenced by the old one. Is that expected behavior ?

Reply:
> Does order matter? Maybe we can use chunked_hash_map here

Nope order doesn't matter, let me switch and see what happens

> t is unclear to me how we handle producer with bupped epoch. It seems that we only forget about producer when all transactions are aborted. The producer with new epoch will always be fenced by the old one. Is that expected behavior ?

This happens in begin_tx()..  this check..


```
    } else if (pid.epoch > current_pid.epoch) {
        // abort any transactions from the older instance of the producer.
        auto ar = co_await do_abort_tx(
          synced_term, producer, std::nullopt, _sync_timeout);
        if (ar != tx_errc::none) {
            vlog(
              _ctx_log.trace,
              "can't begin tx {} because abort of a prev tx {} failed with {}",
              pid,
              current_pid,
              ar);
            co_return tx_errc::stale;
        }
    
```

This ensures that any in progress transaction with older epoch are "aborted", which ensures the _transaction_state is reset accordingly.. then we ensure there are no transactions are in progress before we apply a new fence batch..

```
    vassert(
      !producer->has_transaction_in_progress(),
      "There is an in progress transaction for producer, invalid state: {}",
      producer);

```

then in the producer state we update the latest epoch for the producer..

```
void producer_state::apply_transaction_begin(
  const model::record_batch_header& header,
  const fence_batch_data& parsed_batch) {
    vassert(
      !has_transaction_in_progress() && !_active_transaction_hook.is_linked(),
      "Transaction already in progress {} for producer {}, hook {}",
      _transaction_state,
      *this,
      _active_transaction_hook.is_linked());
    // update pid incase the producer got fenced.
    const auto& pid = parsed_batch.bid.pid;
    vassert(
      pid.epoch >= _id.epoch,
      "[{}] Invalid fencing using a pid with lower epoch: {}",
      *this,
      header);
    _id = pid;
````